libjulia/optimiser/ASTCopier.cpp

```diff
@@ -0,0 +1,174 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * Creates an independent copy of an AST, renaming identifiers to be unique.
+ */
+
+#include <libjulia/optimiser/ASTCopier.h>
+
+#include <libsolidity/inlineasm/AsmData.h>
+
+#include <libsolidity/interface/Exceptions.h>
+
+#include <libdevcore/Common.h>
+
+using namespace std;
+using namespace dev;
+using namespace dev::julia;
+
+
+Statement ASTCopier::operator()(Instruction const& _instruction)
+{
+	return _instruction;
```

Comment:
If `StackAssignment` and `Label` is disabled below, then this should be disabled too. Though At this point this isn't pure Julia rather both versions and I'd argue we support all AST elements, unless you want to add a boolean for choosing between julia and non-julia mode.

Reply:
This component does not validate the input, the assumption is that everything went through the proper analyzers. So it works for both flavours. Also, `Instruction` is part of `FunctionalInstruction`. Perhaps we should change that globally in a later PR, so that `FunctionalInstruction` directly contains the instruction enum value and what is currently `Instruction` is only used for non-functional instructions (which will be deprecated later).

Reply:
Ah, just re-read your comment: The optimizer does not work for stackassignment because it breaks some of the assumptions.

Reply:
`StackAssignment` and `Label` have an assertion. It means one cannot use `ASTCopier` on non-julia inline assembly.

Reply:
> Perhaps we should change that globally in a later PR, so that FunctionalInstruction directly contains the instruction enum value

That is a good idea. There seems to be no reason to have it the way it is - it has now to source locations too.